src/GraphQL/Types/Scalars/ShortGraphType.cs

```diff
@@ -12,25 +12,56 @@ public class ShortGraphType : ScalarGraphType
         /// <inheritdoc/>
         public override object ParseLiteral(IValue value) => value switch
         {
-            IntValue intValue => short.MinValue <= intValue.Value && intValue.Value <= short.MaxValue ? (short?)intValue.Value : null,
-            LongValue longValue => short.MinValue <= longValue.Value && longValue.Value <= short.MaxValue ? (short?)longValue.Value : null,
-            BigIntValue bigIntValue => short.MinValue <= bigIntValue.Value && bigIntValue.Value <= short.MaxValue ? (short?)bigIntValue.Value : null,
-            _ => null
+            IntValue intValue => checked((short)intValue.Value),
+            LongValue longValue => checked((short)longValue.Value),
+            BigIntValue bigIntValue => checked((short)bigIntValue.Value),
+            NullValue _ => null,
+            _ => ThrowLiteralConversionError(value)
         };
 
-        /// <inheritdoc/>
-        public override object ParseValue(object value) => ValueConverter.ConvertTo(value, typeof(short));
-
         /// <inheritdoc/>
         public override bool CanParseLiteral(IValue value) => value switch
         {
             IntValue intValue => short.MinValue <= intValue.Value && intValue.Value <= short.MaxValue,
             LongValue longValue => short.MinValue <= longValue.Value && longValue.Value <= short.MaxValue,
             BigIntValue bigIntValue => short.MinValue <= bigIntValue.Value && bigIntValue.Value <= short.MaxValue,
+            NullValue _ => true,
             _ => false
         };
 
         /// <inheritdoc/>
-        public override IValue ToAST(object value) => new IntValue(Convert.ToInt16(value));
+        public override object ParseValue(object value) => value switch
+        {
+            short _ => value,
+            null => null,
+            int i => checked((short)i),
+            sbyte sb => checked((short)sb),
+            byte b => checked((short)b),
+            ushort us => checked((short)us),
+            uint ui => checked((short)ui),
+            long l => checked((short)l),
+            ulong ul => checked((short)ul),
+            BigInteger bi => (short)bi,
+            float f => checked((short)f),
+            double d => checked((short)d),
+            decimal d => checked((short)d),
+            _ => ThrowValueConversionError(value)
+        };
+
+        /// <inheritdoc/>
+        public override object Serialize(object value) => value switch
```

Comment:
Why copy-paste this from `ParseValue` that is called by default from `ScalarGraphType.Serialize` ? Also `ParseValue` has cases for floating point numbers.

Reply:
Because within the resolver, you should be returning values of the correct type.  The only reason floating-point support was added for ParseValue is because of the fact that json deserializers may not deserialize values to the correct native .NET type.

Reply:
> Because within the resolver, you should be returning values of the **correct** type. 

What do you mean ? Is it correct to return `(int)1` from resolver for `Short` ? Resolver returns internal representation of the scalar. Then this value converted to the external representation.

Reply:
I meant that you should be returning an integer type for integers.

Reply:
We can be more explicit in our support for internally-represented values, only accepting `int` for `IntGraphType`, and `sbyte` for `SByteGraphType`.  At this point, the added flexibility is fully overflow-checked and does not result in loss of precision.  So I see it as a "bonus feature".

Reply:
"Bonus feature" is good but let's separate integers from floating point numbers. This is a line crossing which we take risks. Remember about `DateGraphType` and time component. Let's be consistent.

Reply:
Then, most likely, we can remove floating-point conversions from `int` and smaller.  `uint`, `long`, and `ulong` probably need floating-point conversions for newtonsoft's json converter to work properly.  Would you agree?

Reply:
> uint, long, and ulong probably need floating-point conversions for newtonsoft's json converter to work properly

Why? Let's test it.

Reply:
Are you talking about that rather big number from JSON string will be deserialized by Newtonsoft/System.Text.Json not as `int/long/ulong/BigInt`, but always as `double` so we de facto always need conversion from `double` to integer scalars ?

Reply:
Removed duplicate code as requested.